### PR TITLE
"Seal" L1T unpacking plugin declarations.

### PR DESCRIPTION
**Pull request by matz-e**
Factor out headers for the (un)packing of the L1T and seal the modules in the `plugins` directory.  Allows other packages to depend on `L1TRawToDigi` as a library, as needed for DQM.

### CODE REVIEW

**Comment by thomreis** [comment]
Can someone launch the tests for this please?

**Comment by dmitrijus** [comment]
+1

**Comment by mdhildreth** [comment]
please test

**Comment by davidlange6** [comment]
@matz-e - what is needed instead is to move things that are plugins into the plugins directory. Perhaps this is everything aside from AMC* and Block (note that the include guard of Block.h needs to be changed to include the subdirectory and package name before anyone can safely depend on this package.

**Comment by matz-e** [comment]
@mulhearn can we get this merged to avoid merge conflicts, please?

**Comment by rekovic** [comment]
@matz-e  I don't see any new commits made after the comment of @davidlange6 above. If not, are those comments intended to be addressed? (same for #15388).

**Comment by matz-e** [comment]
Ah, sorry, didn't see that over the buildbot. Will take care of that.

**Comment by matz-e** [comment]
Stupid question: can we keep the subdirectory structure in `plugins/`? Having the code separated that way was one motivation to keep that in `src/`.

**Comment by rekovic** [comment]
@matz-e.  I think that @davidlange6 can make a final comment, but I don't see why not.

**Comment by matz-e** [comment]
@davidlange6 I've moved all the plugins. Do I need to move the plugin definitions and factories, too? IMHO, they fit better into the `interface`, `src` directories…

@thomreis will this still work with your DQM code?

**Comment by thomreis** [comment]
@matz-e I will check but it should still work.

**Comment by davidlange6** [comment]
plugin definitions need to be in the plugins area. checking what you mean by factories in this context

**Comment by davidlange6** [comment]
things like PackerFactory would need to move, yes

**Comment by thomreis** [comment]
I'm getting this runtime error for my DQM code when pulling in your latest changes.

```
----- Begin Fatal Exception 02-Sep-2016 19:18:55 CEST-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Processing run: 277981 lumi: 140 event: 106464574
   [1] Running path 'rawToDigiPath'
   [2] Calling event method for module L1TRawToDigi/'bmtfDigis'
Exception Message:
unable to load /afs/cern.ch/work/t/treis/ugmt/dqm/clean/test/CMSSW_8_0_16/lib/slc6_amd64_gcc530/poisoned/plugin-poisoned-EventFilterL1TRawToDigiAuto.so because /afs/cern.ch/work/t/treis/ugmt/dqm/clean/test/CMSSW_8_0_16/lib/slc6_amd64_gcc530/poisoned/plugin-poisoned-EventFilterL1TRawToDigiAuto.so: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------
```

**Comment by thomreis** [comment]
To clarify: The runtime error actually happened when I tested the 8_0_x version. So this should rather be in PR #15388

**Comment by matz-e** [comment]
Thanks, I'll see if I can dig out my own testing config again and reproduce.

**Comment by matz-e** [comment]
@davidlange6 I've moved every mention of `EDM_` macros into `plugins` now.

**Comment by matz-e** [comment]
@thomreis with this release, my test setup works…

**Comment by thomreis** [comment]
@matz-e are you going to apply the latest commit also for the 80x version?

**Comment by matz-e** [comment]
@thomreis I'm in the process… already cherry-picked and fixed compilation. Just want to run my test before I push.

**Comment by matz-e** [comment]
@thomreis actually, I had pushed it to test it… m(

My test comes out clean, if yours fails, could you point me to your setup so that I could investigate, please?

**Comment by matz-e** [comment]
FWIW, this should be the last change to this PR (I hope).

**Comment by thomreis** [comment]
@matz-e now that PR #15745 is merged we have conflicts. I guess we should also wait for the other two L1 PR to be merged as I fear they would break the mergeability again even if it is fixed now.

**Comment by thomreis** [comment]
@matz-e the last of the three L1 PRs is merged now. I think we can go ahead and resolve the conflicts that they introduced.

**Comment by matz-e** [comment]
There — that was a lot less painful than I anticipated. @thomreis could you please test this?

**Comment by mmusich** [comment]
this needs a rebase...

**Comment by thomreis** [comment]
The IntermediateMuonPacker.cc/IntermediateMuonUnpacker.cc in /EventFilter/L1TRawToDigi/src/implementations_stage2 should be moved as well. These were new in the L1 PR.

**Comment by matz-e** [comment]
Hm… I see. @mmusich should I force-push a rebase or create a new PR?

**Comment by mmusich** [comment]
if you force-push a rebase you won't loose the branch.

**Comment by matz-e** [comment]
I didn't understand that entirely — I just created a reference branch and forced pushed this.

**Comment by mmusich** [comment]
looks good

**Comment by mmusich** [comment]
please test

**Comment by matz-e** [comment]
Sorry, forgot to include those headers :(

**Comment by mmusich** [comment]
please test
(as a side not, after the rebase, there is no need for alca signature, just asking tests to speedup integration)

**Comment by matz-e** [comment]
Thanks!

**Comment by thomreis** [comment]
Looks like something got messed up in https://github.com/matz-e/cmssw/blob/seal_81x/EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFSetup.cc

Compared to the 8_1_X version:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/EventFilter/L1TRawToDigi/src/implementations_stage2/BMTFSetup.cc

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFSetup.cc [hunk 1]
```diff
@@ -50,9 +50,6 @@ namespace l1t {
       UnpackerMap
       BMTFSetup::getUnpackers(int fed, int board, int amc, unsigned int fw)
       {
-         auto outputMuon = UnpackerFactory::get()->make("stage2::BMTFUnpackerOutput");
-         auto inputMuons = UnpackerFactory::get()->make("stage2::BMTFUnpackerInputs");
-
          UnpackerMap res;
```
Comment (thomreis) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFSetup.cc]: @matz-e there is still a duplication here. The UnpackerMap variable declaration and the for-loop appear again in line 63.

Comment (matz-e) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFSetup.cc]: Indeed.

**Comment by thomreis** [comment]
Can someone start the tests for this please?

**Comment by matz-e** [comment]
FWIW, I successfully ran the `runTheMatrix` workflow that failed above. I was missing a good test workflow before, and will use that in the future.

**Comment by mmusich** [comment]
please test

**Comment by thomreis** [comment]
@rekovic @mulhearn could you check and sign this off please so that it can be merged?

**Comment by rekovic** [comment]
+1

**Comment by thomreis** [comment]
@davidlange6 - can this be merged? It misses only ORP approval for a week.
Also PR #15388 lacks only the ORP +1.

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
@matz-e -seems this just became unmergable - can you rebase it? (just as I thought all the L1 changes were past us)

**Comment by matz-e** [comment]
I'll rebase it (and it's companion for 80X) — I too had hoped we were past L1 changes.

**Comment by matz-e** [comment]
@davidlange6 there we go… I ran `runTheMatrix.py -s --useInput all` and it came back w/o failures.

**Comment by matz-e** [comment]
Can we get this re-tested and merged, please (unless there are some comments that I overlooked)?

**Comment by rekovic** [comment]
+1

**Pull request merged**